src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSpatialDatatype.java

```diff
@@ -184,23 +203,544 @@ protected void serializeToWkb(boolean excludeZMFromWKB, SQLServerSpatialDatatype
+                        addMultiPointToBuffer(buf, 0);
+                    } else {
+                        addMultiPointToBuffer(buf, calculateNumFiguresInThisShape(true));
+                    }
+                    currentWKBShapeIndex++;
+                    break;
+                case MULTILINESTRING:
+                    if (shapes[currentWKBShapeIndex].getFigureOffset() == -1) {
+                        addMultiLineStringToBuffer(buf, 0);
+                    } else {
+                        addMultiLineStringToBuffer(buf, calculateNumFiguresInThisShape(true));
+                    }
+                    currentWKBShapeIndex++;
+                    break;
+                case MULTIPOLYGON:
+                    /*
+                     * increment WKBShapeIndex for all shapes except for GeometryCollection and Multipolygon, since
+                     * their shape index was incremented earlier to be used for calculation.
+                     */
+                    if (shapes[currentWKBShapeIndex].getFigureOffset() == -1) {
+                        addMultiPolygonToBuffer(buf, 0);
+                    } else {
+                        addMultiPolygonToBuffer(buf, calculateNumShapesInThisMultiPolygon());
+                    }
+                    break;
+                case GEOMETRYCOLLECTION:
+                    if (shapes[currentWKBShapeIndex].getFigureOffset() == -1) {
+                        addGeometryCollectionToBuffer(buf, 0);
+                    } else {
+                        addGeometryCollectionToBuffer(buf, calculateNumShapesInThisGeometryCollection());
+                    }
+                    break;
+                case CIRCULARSTRING:
+                    if (shapes[currentWKBShapeIndex].getFigureOffset() == -1) {
+                        addCircularStringToBuffer(buf, 0);
+                    } else {
+                        addCircularStringToBuffer(buf, calculateNumPointsInThisFigure());
+                    }
+                    currentWKBShapeIndex++;
+                    break;
+                case COMPOUNDCURVE:
+                    if (shapes[currentWKBShapeIndex].getFigureOffset() == -1) {
+                        addCompoundCurveToBuffer(buf, 0);
+                    } else {
+                        addCompoundCurveToBuffer(buf, calculateNumCurvesInThisFigure());
+                    }
+                    currentWKBShapeIndex++;
+                    break;
+                case CURVEPOLYGON:
+                    if (shapes[currentWKBShapeIndex].getFigureOffset() == -1) {
+                        addCurvePolygonToBuffer(buf, 0);
+                    } else {
+                        addCurvePolygonToBuffer(buf, calculateNumFiguresInThisShape(false));
+                    }
+                    currentWKBShapeIndex++;
+                    break;
+                default:
+                    // Shouldn't be possible to come here, existing Geometry/Geography doesn't support other cases
```

Comment:
Same here and in a few other default cases below, instead of ignoring unexpected values, we should throw and exception.